### PR TITLE
Cremation animation + refactors and improvements

### PR DESCRIPTION
**Pull request by alsoandanswer**
refactors crematoriums to reuse morgue code because there was a lot of copy pasta. should run a little more lean now.

anywho, here's a video showing the business side of things.

https://user-images.githubusercontent.com/42310821/103446397-0d8e7500-4cba-11eb-8deb-2040993e5a51.mp4

### CODE REVIEW

**Comment by alsoandanswer** [comment]
!review

File: code/_helpers/matrices.dm [hunk 1]
```diff
@@ -25,6 +25,12 @@
 		//doesn't have an object argument because this is "Stacking" with the animate call above
 		//3 billion% intentional
 
+/atom/proc/shake_animation(var/intensity = 8)
+	var/init_px = pixel_x
+	var/shake_dir = pick(-1, 1)
+	animate(src, transform=turn(matrix(), intensity*shake_dir), pixel_x=init_px + 2*shake_dir, time=1)
+	animate(transform=null, pixel_x=init_px, time=6, easing=ELASTIC_EASING)
```
Comment (Geevies) [inline code/_helpers/matrices.dm:32]: this doesn't have the return value the proc it's replacing had. iirc something else in the code relied on it

File: code/game/objects/structures/morgue.dm [hunk 2]
```diff
@@ -29,10 +31,18 @@
...
 		icon_state = "morgue0"
 	else
-		if(contents.len)
-			icon_state = "morgue2"
-		else
-			icon_state = "morgue1"
+		icon_state = "morgue1"
+		var/list/searching = GetAllContents(searchDepth = 3, checkClient = 0) // Search inside bodybags as well.
+		for(var/mob/living/M in searching)
+			if(iscarbon(M))
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:37]: this should consider fakedeath as well

File: code/game/objects/structures/morgue.dm [hunk 3]
```diff
@@ -29,10 +31,18 @@
...
+				var/mob/living/carbon/C = M
+				switch(C.stat)
+					if(DEAD)
+						icon_state = "morgue2"
+					if(UNCONSCIOUS)
+						icon_state = "morgue3"
+					if(CONSCIOUS)
+						icon_state = "morgue4"
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:45]: you can either break here after finding the first mob and setting the state, or you can save the stats. meaning if there's three dead, but one alive, you show the alive one

Comment (Geevies) [reply code/game/objects/structures/morgue.dm:45]: also if possible you can make these names a bit more clear

File: code/game/objects/structures/morgue.dm [hunk 4]
```diff
@@ -101,21 +89,21 @@
 
 /obj/structure/morgue/attackby(obj/P, mob/user)
 	if(P.ispen())
-		var/t = input(user, "What would you like the label to be?", text("[]", src.name), null)  as text
+		var/t = input(user, "What would you like the label to be?", "[src.name]") as text
 		if(user.get_active_hand() != P)
 			return
-		if((!in_range(src, usr) && src.loc != user))
+		if((!in_range(src, usr) > 1 && src.loc != user))
 			return
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:96]: i dunno if this works with the ! in front, looks messy

File: code/game/objects/structures/morgue.dm [hunk 5]
```diff
@@ -101,21 +89,21 @@
...
 		t = sanitizeSafe(t, MAX_NAME_LEN)
 		if(t)
-			name = "Morgue- '[t]'"
+			name = "initial[name]- '[t]'"
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:99]: [initial(name)]

File: code/game/objects/structures/morgue.dm [hunk 6]
```diff
@@ -101,21 +89,21 @@
...
 		else
-			name = "Morgue"
+			name = "initial[name]"
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:101]:
```suggestion
			name = initial(name)
```

File: code/game/objects/structures/morgue.dm [hunk 7]
```diff
@@ -224,214 +213,146 @@
...
+				for(var/I, I < 60, I++)
+					C.bodytemperature += 50
+					C.adjustFireLoss(20)
+					C.adjustBrainLoss(5)
+
+					if(C.stat == DEAD || !(C in searching)) //In case we die or are removed at any point.
+						cremating = initial(cremating)
+						locked = initial(locked)
+						update_icon()
+						break
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:261]: shouldn't this be `continue`

File: code/game/objects/structures/morgue.dm [hunk 8]
```diff
@@ -224,214 +213,146 @@
...
+			if(M.stat == DEAD)
+				if(round_is_spooky())
+					if(prob(50))
+						playsound(src, 'sound/effects/ghost.ogg', 10, 5)
+					else
+						playsound(src, 'sound/effects/ghost2.ogg', 10, 5)
+
+				admin_attack_log(A, M, "Cremated their victim.", "Was cremated.", "cremated")
+				if(desperation)
+					M.audible_message("<b>[M]'s</b> screams cease, as does any movement within the [src]. All that remains is a dull, empty silence.")
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:302]:
```suggestion
					M.audible_message("<b>[M]'s</b> screams cease, as does any movement within \the [src]. All that remains is a dull, empty silence.")
```

File: code/game/objects/structures/morgue.dm [hunk 9]
```diff
@@ -224,214 +213,146 @@
...
+				elses
+					M.audible_message("The [src] quietly shuts off. All that remains is a dull, empty silence.")
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:304]:
```suggestion
					M.audible_message("\The [src] quietly shuts off. All that remains is a dull, empty silence.")
```

**Review by Geevies — CHANGES_REQUESTED** [review]
did you sprite this?

**Comment by alsoandanswer** [comment]
> 
> 
> did you sprite this?

i got these from bay, they don't deviate too much from our current morgues anyway so i don't see any problems. they're credited in the changelog.

i also added some animations by myself.

File: code/_helpers/matrices.dm [hunk 10]
```diff
@@ -25,6 +25,13 @@
...
+	return
```
Comment (Geevies) [inline code/_helpers/matrices.dm:33]: this still doesn't return the value

File: code/game/objects/structures/morgue.dm [hunk 11]
```diff
@@ -29,10 +31,24 @@
...
+				switch(C.stat)
+					if(DEAD)
+						icon_state = "morgue2"
+						break
+					if(UNCONSCIOUS)
+						icon_state = "morgue3"
+						break
+					if(CONSCIOUS)
+						icon_state = "morgue4"
+						break
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:51]: if you're doing it this way, just break after you find a carbon and do the stuff, you don't need to add it to each line

File: code/game/objects/structures/morgue.dm [hunk 12]
```diff
@@ -101,21 +95,21 @@
 
 /obj/structure/morgue/attackby(obj/P, mob/user)
 	if(P.ispen())
-		var/t = input(user, "What would you like the label to be?", text("[]", src.name), null)  as text
+		var/t = input(user, "What would you like the label to be?", "[src.name]") as text
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:98]:
```suggestion
		var/t = input(user, "What would you like the label to be?", name) as text
```

File: code/game/objects/structures/morgue.dm (hunk 4 above)
Comment (Geevies) [reply code/game/objects/structures/morgue.dm:102]: you resolved this one without actually changing it

File: code/game/objects/structures/morgue.dm [hunk 13]
```diff
@@ -224,214 +219,147 @@
...
+		cremating = initial(cremating)
+		locked = initial(locked)
+		update_icon()
+		return
 	else
-		visible_message(SPAN_WARNING("You hear a roar as \the [src] activates."))
-		cremating = TRUE
-		locked = TRUE
+		if(length(search_contents_for(/obj/item/disk/nuclear)))
+			to_chat(loc, "The button's status indicator flashes yellow, indicating that something important is inside the crematorium, and must be removed.")
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:246]: does `to_chat(loc` work?

File: code/game/objects/structures/morgue.dm [hunk 14]
```diff
@@ -224,214 +219,147 @@
...
+			return
+		src.audible_message(SPAN_WARNING("You hear a roar as the [src] activates."), 1)
```
Comment (Geevies) [inline code/game/objects/structures/morgue.dm:248]:
```suggestion
		src.audible_message(SPAN_WARNING("You hear a roar as \the [src] activates."), 1)
```

**Review by Geevies — CHANGES_REQUESTED** [review]

**Comment by Geevies** [comment]
remember to say that you ported something in the PR description

File: code/game/objects/structures/morgue.dm (hunk 4 above)
Comment (alsoandanswer) [reply code/game/objects/structures/morgue.dm:102]: yes, it works

**Review by NonQueueingMatt — APPROVED** [review]

**Review by fernerr — APPROVED** [review]

**Pull request merged**